File: lib/error-extractor.ts
/**
 * Extracted error information from Jira ticket
 */
export interface ExtractedErrorInfo {
  errorMessage?: string;
  stackTrace?: string;
  testFailure?: string;
  errorType?: string;
  lineNumber?: number;
  filePath?: string;
}

/**
 * Extracts error information from Jira ticket description
 */
export function extractErrorInfo(description: string): ExtractedErrorInfo {
  const info: ExtractedErrorInfo = {};

  // Extract stack traces (common patterns)
  const stackTracePatterns = [
    /(?:Error|Exception|TypeError|ReferenceError|SyntaxError)[:\s]+([^\n]+)/i,
    /at\s+([^\s]+)\s+\(([^:]+):(\d+):(\d+)\)/g,
    /(?:Stack trace|StackTrace):\s*([\s\S]+?)(?:\n\n|\n[A-Z]|$)/i,
  ];

  for (const pattern of stackTracePatterns) {
    const matches = description.match(pattern);
    if (matches) {
      if (!info.stackTrace) {
        info.stackTrace = '';
      }
      info.stackTrace += matches[0] + '\n';
    }
  }

  // Extract error messages (common patterns)
  const errorMessagePatterns = [
    /(?:Error|Exception|Failed|Fails?):\s*([^\n]+)/i,
    /(?:Error message|Error Message):\s*([^\n]+)/i,
    /"([^"]*Error[^"]*)"|'([^']*Error[^']*)'/i,
  ];

  for (const pattern of errorMessagePatterns) {
    const match = description.match(pattern);
    if (match && !info.errorMessage) {
      info.errorMessage = match[1] || match[2] || match[0];
      break;
    }
  }

  // Extract test failures
  const testFailurePatterns = [
    /(?:Test|Spec)\s+(?:failed|failure|error)[:\s]+([^\n]+)/i,
    /(?:FAIL|FAILED|ERROR)\s+([^\n]+)/i,
    /(?:Expected|Expected:)\s+([^\n]+)/i,
    /(?:Actual|Actual:)\s+([^\n]+)/i,
  ];

  for (const pattern of testFailurePatterns) {
    const match = description.match(pattern);
    if (match && !info.testFailure) {
      info.testFailure = match[1] || match[0];
      break;
    }
  }

  // Extract file path and line number from stack traces
  const fileLinePattern = /([\/\w\-\.]+\.(?:ts|tsx|js|jsx)):(\d+)/;
  const fileMatch = description.match(fileLinePattern);
  if (fileMatch) {
    info.filePath = fileMatch[1];
    info.lineNumber = parseInt(fileMatch[2], 10);
  }

  // Extract error type
  const errorTypePattern = /(TypeError|ReferenceError|SyntaxError|Error|Exception|ValidationError|RuntimeError)/i;
  const typeMatch = description.match(errorTypePattern);
  if (typeMatch) {
    info.errorType = typeMatch[1];
  }

  return info;
}

/**
 * Formats extracted error info for AI prompt
 */
export function formatErrorInfoForPrompt(errorInfo: ExtractedErrorInfo): string {
  if (!errorInfo.errorMessage && !errorInfo.stackTrace && !errorInfo.testFailure) {
    return '';
  }

  let formatted = '\n---\nError Information:\n\n';

  if (errorInfo.errorType) {
    formatted += `Error Type: ${errorInfo.errorType}\n`;
  }

  if (errorInfo.errorMessage) {
    formatted += `Error Message: ${errorInfo.errorMessage}\n`;
  }

  if (errorInfo.filePath && errorInfo.lineNumber) {
    formatted += `Location: ${errorInfo.filePath}:${errorInfo.lineNumber}\n`;
  }

  if (errorInfo.stackTrace) {
    formatted += `\nStack Trace:\n${errorInfo.stackTrace}\n`;
  }

  if (errorInfo.testFailure) {
    formatted += `\nTest Failure:\n${errorInfo.testFailure}\n`;
  }

  formatted += '---\n';

  return formatted;
}

File: src/main/java/com/automata/service/ErrorExtractor.java
package com.automata.service;

import com.automata.model.ExtractedErrorInfo;
import org.springframework.stereotype.Service;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class ErrorExtractor {

    public ExtractedErrorInfo extractErrorInfo(String description) {
        if (description == null || description.trim().isEmpty()) {
            return new ExtractedErrorInfo();
        }

        ExtractedErrorInfo info = new ExtractedErrorInfo();

        // 1. Extract Stack Traces
        Pattern[] stackTracePatterns = {
            Pattern.compile("(?:Error|Exception|TypeError|ReferenceError|SyntaxError)[:\\s]+([^\\n]+)", Pattern.CASE_INSENSITIVE),
            Pattern.compile("at\\s+([^\\s]+)\\s+\\(([^:]+):(\\d+):(\\d+)\\)", Pattern.CASE_INSENSITIVE),
            Pattern.compile("(?:Stack trace|StackTrace):\\s*([\\s\\S]+?)(?:\\n\\n|\\n[A-Z]|$)", Pattern.CASE_INSENSITIVE)
        };

        StringBuilder stackTraceBuilder = new StringBuilder();
        for (Pattern pattern : stackTracePatterns) {
            Matcher matcher = pattern.matcher(description);
            if (matcher.find()) {
                stackTraceBuilder.append(matcher.group(0)).append("\n");
            }
        }
        if (stackTraceBuilder.length() > 0) {
            info.setStackTrace(stackTraceBuilder.toString().trim());
        }

        // 2. Extract Error Messages
        Pattern[] errorMessagePatterns = {
            Pattern.compile("(?:Error|Exception|Failed|Fails?):\\s*([^\\n]+)", Pattern.CASE_INSENSITIVE),
            Pattern.compile("(?:Error message|Error Message):\\s*([^\\n]+)", Pattern.CASE_INSENSITIVE),
            Pattern.compile("\"([^\"]*Error[^\"]*)\"|'([^']*Error[^']*)'", Pattern.CASE_INSENSITIVE)
        };

        for (Pattern pattern : errorMessagePatterns) {
            Matcher matcher = pattern.matcher(description);
            if (matcher.find()) {
                String match = matcher.group(1) != null ? matcher.group(1) : 
                              (matcher.group(2) != null ? matcher.group(2) : matcher.group(0));
                info.setErrorMessage(match);
                break;
            }
        }

        // 3. Extract Test Failures
        Pattern[] testFailurePatterns = {
            Pattern.compile("(?:Test|Spec)\\s+(?:failed|failure|error)[:\\s]+([^\\n]+)", Pattern.CASE_INSENSITIVE),
            Pattern.compile("(?:FAIL|FAILED|ERROR)\\s+([^\\n]+)", Pattern.CASE_INSENSITIVE),
            Pattern.compile("(?:Expected|Expected:)\\s+([^\\n]+)", Pattern.CASE_INSENSITIVE),
            Pattern.compile("(?:Actual|Actual:)\\s+([^\\n]+)", Pattern.CASE_INSENSITIVE)
        };

        for (Pattern pattern : testFailurePatterns) {
            Matcher matcher = pattern.matcher(description);
            if (matcher.find()) {
                String match = matcher.group(1) != null ? matcher.group(1) : matcher.group(0);
                info.setTestFailure(match);
                break;
            }
        }

        // 4. Extract File Path and Line Number
        Pattern fileLinePattern = Pattern.compile("([\\/\\w\\-\\.]+\\.(?:ts|tsx|js|jsx)):(\\d+)");
        Matcher fileLineMatcher = fileLinePattern.matcher(description);
        if (fileLineMatcher.find()) {
            info.setFilePath(fileLineMatcher.group(1));
            try {
                info.setLineNumber(Integer.parseInt(fileLineMatcher.group(2)));
            } catch (NumberFormatException ignored) {}
        }

        // 5. Extract Error Type
        Pattern errorTypePattern = Pattern.compile("(TypeError|ReferenceError|SyntaxError|Error|Exception|ValidationError|RuntimeError)", Pattern.CASE_INSENSITIVE);
        Matcher errorTypeMatcher = errorTypePattern.matcher(description);
        if (errorTypeMatcher.find()) {
            info.setErrorType(errorTypeMatcher.group(1));
        }

        return info;
    }

    public String formatErrorInfoForPrompt(ExtractedErrorInfo errorInfo) {
        if (errorInfo == null || (errorInfo.getErrorMessage() == null && errorInfo.getStackTrace() == null && errorInfo.getTestFailure() == null)) {
            return "";
        }

        StringBuilder builder = new StringBuilder();
        builder.append("\n---\nError Information:\n\n");

        if (errorInfo.getErrorType() != null) {
            builder.append("Error Type: ").append(errorInfo.getErrorType()).append("\n");
        }
        if (errorInfo.getErrorMessage() != null) {
            builder.append("Error Message: ").append(errorInfo.getErrorMessage()).append("\n");
        }
        if (errorInfo.getFilePath() != null && errorInfo.getLineNumber() != null) {
            builder.append("Location: ").append(errorInfo.getFilePath()).append(":").append(errorInfo.getLineNumber()).append("\n");
        }
        if (errorInfo.getStackTrace() != null) {
            builder.append("\nStack Trace:\n").append(errorInfo.getStackTrace()).append("\n");
        }
        if (errorInfo.getTestFailure() != null) {
            builder.append("\nTest Failure:\n").append(errorInfo.getTestFailure()).append("\n");
        }

        builder.append("---\n");
        return builder.toString();
    }
}
